complex/cargo/list/src/lib.rs

```diff
@@ -37,228 +49,778 @@ pub struct ListManager {
 }
 
 impl ListManager {
-    pub fn new(store: Arc<Store>) -> ListManager {
-        let manager = ListManager {
+    pub fn new(store: Arc<Store>) -> Result<ListManager, list_errors::Error> {
+        let mut manager = ListManager {
             store: store,
         };
-        manager.create_labels_table();
-        manager.create_items_table();
-        manager.create_item_labels_table();
-        manager
+        manager.create_labels_table()?;
+        manager.create_items_table()?;
+        Ok(manager)
     }
 
-    pub fn create_labels_table(&self) {
-        let sql = r#"CREATE TABLE IF NOT EXISTS labels (
-                name TEXT NOT NULL PRIMARY KEY,
-                color TEXT NOT NULL
-            )"#;
-        let db = self.store.write_connection();
-        db.execute(sql, &[]).unwrap();
+    fn write_connection(&mut self) -> &mut Store {
+        Arc::get_mut(&mut self.store).unwrap()
     }
 
-    pub fn create_label(&self, name: String, color: String) -> Option<Label> {
-        let sql = r#"INSERT INTO labels (name, color) VALUES (?1, ?2)"#;
-        let db = self.store.write_connection();
-        db.execute(sql, &[&name, &color]).unwrap();
-        self.fetch_label(&name)
+    pub fn create_labels_table(&mut self) -> Result<(), list_errors::Error> {
+        let schema = r#"[{  :db/ident     :label/name
+    :db/valueType :db.type/string
+    :db/cardinality :db.cardinality/one
+    :db/unique :db.unique/identity
+    :db/fulltext true },
+ {  :db/ident     :label/color
+    :db/valueType :db.type/string
+    :db/cardinality :db.cardinality/one }]"#;
+        let _ = self.write_connection().transact(schema)?;
+        Ok(())
     }
 
-    pub fn fetch_label(&self, name: &String) -> Option<Label> {
-        let sql = r#"SELECT name, color FROM labels WHERE name=?"#;
+    pub fn create_label(&mut self, name: String, color: String) -> Result<Option<Label>, list_errors::Error> {
+        let query = format!("[{{ :label/name \"{0}\" :label/color \"{1}\" }}]", &name, &color);
+        self.write_connection().transact(&query)?;
+        self.fetch_label(&name)
+    }
 
-        let conn = self.store.read_connection();
-        let mut stmt = conn.prepare(sql).unwrap();
-        let mut label_iter = stmt.query_map(&[name], |row| {
-            Label {
-                name: row.get(0),
-                color: row.get(1)
+    pub fn fetch_label(&self, name: &String) -> Result<Option<Label>, list_errors::Error> {
+        let query = r#"[:find ?eid, ?name, ?color
+            :in ?name
+            :where
+            [?eid :label/name ?name]
+            [?eid :label/color ?color]
+        ]"#;
+        let result = Arc::clone(&self.store).query_args(query, &[&(&"?name".to_string(), &name)])?;
```

Comment:
Why do you need to clone the `Arc` here?

And you might want to revisit that `query_args` signature. This is what we do in the query tests:

```
    let inputs = QueryInputs::with_value_sequence(vec![(Variable::from_valid_name("?a"), TypedValue::Ref(a))]);
```

Reply:
I thought you had to clone the `Arc` in order to create another reference to it? Maybe I have misunderstood `Arc`?

Reply:
With the query_arg signature, I was trying to avoid having to deal with `Variable` and `TypedValue` from crates external to `Store` and providing traits within store to perform that transformation. This felt cleaner to me, hiding the Mentat implementation details from the calling code.

Reply:
No, you didn't misunderstand — in order to create a new reference, you need to clone it. But here I don't think you need a new reference to the store — you are trying to _dereference_ the existing reference so you can call a method on it, no?

Re args and such: I think we should expose `TypedValue` to calling code. It's the simplest approach. We can also _serialize_ (e.g., to JSON or by exposing traits that allow `into`), but it should be possible to get things from the horse's mouth.